### ISSUE TITLE
release-3.1.0: internal/server/server.go:19:12: pattern out/*: no matching files found

### ISSUE THREAD

**Issue by jlsalvador** [TimothyYe/godns#231]
**Describe the bug**  
You _forget_ the "out" directory for the release 3.1.0.

The makefile "build" does not compiles the web, the "out" directory. What do you think about add a dependency to a new target? Maybe a web?

The new target could execute:
- cd web (change workdir)
- npm ci
- npm run build
- mv out ../internal/server/

Another alternative could be to write an `embed.go` in the project root, in order to include the folder "web/out".

Thoughts?

**The frequency of this issue**  
- [x] This issue is easy to reproduce.
- [ ] This issue is not easy to reproduce.

**Steps to Reproduce**  
Download latest release, try to compile it.

**Your GoDNS configuration**  
-

**Expected behavior**  
A successful 3.1.0 binary.

**Screenshots**  
-

**Environment (please complete the following information):**  
 - Compiling the AUR package in a clean working directory.

**Additional context**  
- Time for a 3.1.1 release. 👍

**Comment by TimothyYe** [comment]
Thanks for the quick feedback, I've added one line of annotation at the beginning of the file `internal/server/server.go` to copy the `out` directory generated by the frontend project.  So theoretically, you need to compile & build the frontend project first, and run `go generate ./...`, it copies the `out` directory to `internal/server`.

Before doing that, you need to set up the frontend development environment first, please refer to: https://github.com/TimothyYe/godns?tab=readme-ov-file#setup-the-frontend-development-environment

I'll update the `Makefile` in `3.1.1` release.

**Comment by diizzyy** [comment]
I'm also seeing this in 3.1.2

**Comment by TimothyYe** [comment]
@jlsalvador @diizzyy 

I'll update the Makefile with two available building commands:

```bash
make all #builds all the projects, including the frontend project & the backend project
make build # only build the backend project
```

**Comment by jlsalvador** [comment]
> @jlsalvador @diizzyy
> 
> I'll update the Makefile with two available building commands:
> 
> ```shell
> make all #builds all the projects, including the frontend project & the backend project
> make build # only build the backend project
> ```

I was working on it a bit. https://github.com/jlsalvador/godns/commit/0765d1054c7922772e004f33bed28d9c826380d9
Take a look for improvement. 👍

**Comment by TimothyYe** [comment]
@diizzyy If you'd like to build it with your scripts, here is the workaround:

```bash
mkdir -p ./internal/server/out
touch ./internal/server/out/index.html
GO111MODULE=on go build cmd/godns/godns.go
```

**Comment by diizzyy** [comment]
@TimothyYe 
Thanks, it's actually for the FreeBSD port ( https://repology.org/project/godns/versions ) and that workaround seems to do the trick.

**Comment by TimothyYe** [comment]
@diizzyy Yes, it depends on whether you want to include the frontend project at compile time.

**Comment by diizzyy** [comment]
There's a strict rule that no internet access is allowed during builds (which makes sense) however that makes npm more or less a no go (or a major pain I'm not willing to spend time on) so unless we can get a precompiled frontend there's not going to be packaged version with the frontend.

**Comment by TimothyYe** [comment]
@diizzyy Thanks for the explanation, however, I still have a bit of confusion. If there is no internet connection during builds, when compiling this Go project, if new modules are introduced in the code, they will also be downloaded from the internet during the compilation period. How is this achieved?

**Comment by diizzyy** [comment]
By build I guess I should say compilation as packaging is divided into different stages ( https://docs.freebsd.org/en/books/porters-handbook/book/#options-targets ).

See also:
https://docs.freebsd.org/en/books/porters-handbook/book/#using-go
https://cgit.freebsd.org/ports/tree/Mk/Uses/go.mk

**Comment by TimothyYe** [comment]
@diizzyy I asked ChatGPT for an example of a Makefile to build the port, which includes npm and Go as dependencies. I'm not sure if it will be helpful:

```bash
# Ports metadata
PORTNAME=       myproject
DISTVERSION=    1.0.0
CATEGORIES=     www

MAINTAINER=     me@example.com
COMMENT=        An example Go project with embedded Next.js static files

LICENSE=        MIT

BUILD_DEPENDS=  npm:www/npm \
                go:lang/go

USES=           go:modules

# Use the post-extract target to install Node.js dependencies and build the Next.js project
post-extract:
    @cd ${WRKSRC}/web && ${SETENV} ${MAKE_ENV} npm install
    @cd ${WRKSRC}/web && ${SETENV} ${MAKE_ENV} npm run build

# Use the pre-build target to move the Next.js build output to the location expected by the Go code
pre-build:
    @${MKDIR} ${WRKSRC}/internal/server
    @${MV} ${WRKSRC}/web/out/* ${WRKSRC}/internal/server

# Use the do-build target to build the Go project, embedding the Next.js static files
do-build:
    @cd ${WRKSRC} && ${SETENV} ${MAKE_ENV} go build -o ${PORTNAME} cmd/${PORTNAME}/main.go

# Use the do-install target to install the built Go binary
do-install:
    ${INSTALL_PROGRAM} ${WRKSRC}/${PORTNAME} ${STAGEDIR}${PREFIX}/bin/${PORTNAME}

.include <bsd.port.mk>
```

Key Points:
* __BUILD_DEPENDS__: This line declares build-time dependencies for your port, including npm for building the Next.js project and go for compiling the Go project.

* __USES__: This directive specifies that the port uses Go modules for managing Go dependencies.

* __post-extract__: This target is run after extracting your port's distfiles but before building. It's used here to install Node.js dependencies and build the Next.js project. Adjust the paths according to your project's structure.

**Comment by diizzyy** [comment]
npm is more complicated than that (which is why you rarely see it packaged in any repo).

See below for example
https://cgit.freebsd.org/ports/tree/textproc/write-good/Makefile
https://cgit.freebsd.org/ports/tree/www/py-django-hijack/Makefile

**Comment by TimothyYe** [comment]
Hi @diizzyy , have you figured out the solution yet?

**Comment by diizzyy** [comment]
Hi,

Regarding packaing for FreeBSD there's only a console/no webui version given that npm (and friends) is pretty much impossible to package. Some projects offers a "pre-built/compiled" web-ui that's shipped with the release archive or separately however not sure if that's a viable option in this case.
